README.md
![tests](https://github.com/hoangnm/react-native-week-view/actions/workflows/tests.yml/badge.svg) ![linter](https://github.com/hoangnm/react-native-week-view/actions/workflows/linter.yml/badge.svg) ![codeql](https://github.com/hoangnm/react-native-week-view/actions/workflows/codeql-analysis.yml/badge.svg)

# react-native-week-view
The week view component for react-native.

* Supported in Android and iOS
* Many user interactions supported: **drag and drop events**, edit events, swipe through pages, event press, grid press, etc
* Customizable styles
* Multiple locale support

Checkout the full [documentation site](https://hoangnm.github.io/react-native-week-view), including [guides](https://hoangnm.github.io/react-native-week-view/category/full-api), [full API](https://hoangnm.github.io/react-native-week-view/category/full-api) and [known issues](https://hoangnm.github.io/react-native-week-view/troubleshoot#known-issues).


![basic-example](webdocs/static/img/weekview-example.gif) ![drag-drop](webdocs/docs/guides/img/drag-drop.gif)


## Installation

> `npm install --save react-native-week-view`

or

> `yarn add react-native-week-view`

**Requirements:** install peer dependencies [react-native-gesture-handler](https://docs.swmansion.com/react-native-gesture-handler/docs/installation/) >= 2.4.1 and [react-native-reanimated](https://docs.swmansion.com/react-native-reanimated/docs/fundamentals/installation) >=2.0.0, which we use to provide smoother interactions and animations (e.g. drag and drop).


## Basic usage

```js
import WeekView from 'react-native-week-view';

const myEvents = [
 {
 id: 1,
 description: 'Event',
 startDate: new Date(2021, 3, 15, 12, 0),
 endDate: new Date(2021, 3, 15, 12, 30),
 color: 'blue',
 // ... more properties if needed,
 },
 // More events...
];

const MyComponent = () => (
 <WeekView
 events={myEvents}
 selectedDate={new Date(2021, 3, 15)}
 numberOfDays={7}
 />
);

```

See the [full documentation](https://hoangnm.github.io/react-native-week-view) with more examples and details!


## Changelog

API is still unstable, minor updates before v1.0.0 can include breaking changes, adhering to Semantic Versioning.
See [CHANGELOG.md](./CHANGELOG.md) for details.

## Contributors

<a href="https://github.com/hoangnm/react-native-week-view/graphs/contributors">
 <img src="https://contrib.rocks/image?repo=hoangnm/react-native-week-view" />
</a>

Made with [contrib.rocks](https://contrib.rocks).


CHANGELOG.md
# Changelog

Minor version updates before v1.0.0 can include breaking changes, which will always be listed here.

## 0.30.0 - 2023-08-27

- Pinch vertical gesture [#316](https://github.com/hoangnm/react-native-week-view/pull/316)
## 0.29.0 - 2023-07-19

- Support all-day events [#306](https://github.com/hoangnm/react-native-week-view/pull/306)
- Add eventTextStyle prop and event.textStyle field [#313](https://github.com/hoangnm/react-native-week-view/pull/313)
## 0.28.4 - 2023-04-17

- Fix UI bug [#305](https://github.com/hoangnm/react-native-week-view/pull/305)
- Add missing props [#312](https://github.com/hoangnm/react-native-week-view/pull/312)
## 0.28.3 - 2023-04-04

- Update peer dependencies [#308](https://github.com/hoangnm/react-native-week-view/pull/308)
## 0.28.2 - 2023-03-30

- Fix overlap-ignore bug [#299](https://github.com/hoangnm/react-native-week-view/pull/299)
## 0.28.1 - 2023-02-26

- Fix ANR: Add runOnJS prop [#297](https://github.com/hoangnm/react-native-week-view/pull/297)

## 0.28.0 - 2022-11-08

- Feature: blocks-like events [#267](https://github.com/hoangnm/react-native-week-view/pull/267)

## 0.27.0 - 2022-10-21

- Expose VirtualizedList props [#280](https://github.com/hoangnm/react-native-week-view/pull/280)

## 0.26.0 - 2022-10-14

- Feature: allow scrolling 1 day at the time [#266](https://github.com/hoangnm/react-native-week-view/pull/266)

## 0.25.1 - 2022-09-30

- Support react version from 16.8 [#273](https://github.com/hoangnm/react-native-week-view/pull/273)

## 0.25.0 - 2022-09-30

- Fix incorrect headerTextStyle type [#270](https://github.com/hoangnm/react-native-week-view/pull/270)
- Add drag-config: afterLongPress [#264](https://github.com/hoangnm/react-native-week-view/pull/264)

## 0.24.0 - 2022-09-24

- Improve TS + ESLint [#265](https://github.com/hoangnm/react-native-week-view/pull/265)
- Add support for stacking events [#255](https://github.com/hoangnm/react-native-week-view/pull/255)

## 0.23.0 - 2022-09-14

- Add typescript support and fix some minor bugs [#262](https://github.com/hoangnm/react-native-week-view/pull/262)
- Fix broken goToDate [#257](https://github.com/hoangnm/react-native-week-view/pull/257)

## 0.22.0 - 2022-09-02

- Add support for styles and disable drag/pass per event [#254](https://github.com/hoangnm/react-native-week-view/pull/254)

## 0.21.0 - 2022-08-30

- Add timesColumnWidth prop [#253](https://github.com/hoangnm/react-native-week-view/pull/253)

## 0.20.0 - 2022-08-09

- Add FlatList optimization props [#250](https://github.com/hoangnm/react-native-week-view/pull/250)

## 0.19.0 - 2022-08-09

- Support height responsiveness [#229](https://github.com/hoangnm/react-native-week-view/pull/229)

## 0.18.0 - 2022-07-21

- Add press event to edit [#227](https://github.com/hoangnm/react-native-week-view/pull/227)

## 0.17.0 - 2022-07-14

- Breaking change: add [react-native-gesture-handler](https://docs.swmansion.com/react-native-gesture-handler/docs/installation/) v2 and [react-native-reanimated](https://docs.swmansion.com/react-native-reanimated/docs/fundamentals/installation) v2 as peer dependencies. This change helps drag&drop smoother [#175](https://github.com/hoangnm/react-native-week-view/pull/175)

## 0.16.0 - 2022-06-04

- Improve performance [#210](https://github.com/hoangnm/react-native-week-view/pull/210)

## 0.15.0 - 2022-05-27

- Add scrollToTimeScrolled callback [#208](https://github.com/hoangnm/react-native-week-view/pull/208)

## 0.14.0 - 2022-05-24

- Add scrollToTime method [#205](https://github.com/hoangnm/react-native-week-view/pull/205)

## 0.13.0 - 2022-04-03

- Support orientation changes [#201](https://github.com/hoangnm/react-native-week-view/pull/201)

## 0.12.0 - 2022-04-03

- Add beginAgendaAt and endAgendaAt props [#183](https://github.com/hoangnm/react-native-week-view/pull/183)

## 0.11.0 - 2022-04-03

- Add DayHeaderComponent prop [#182](https://github.com/hoangnm/react-native-week-view/pull/182)

## 0.10.0 - 2022-03-27

- Add onDayPress and onMonthPress props [#176](https://github.com/hoangnm/react-native-week-view/pull/176)
- Add support for React 17 [#173](https://github.com/hoangnm/react-native-week-view/pull/173)

## 0.9.0 - 2022-01-18

- Add gridRowStyle and gridColumnStyle props [#158](https://github.com/hoangnm/react-native-week-view/pull/158)

## 0.8.1 - 2021-12-30

- Fix dragging event issue [#152](https://github.com/hoangnm/react-native-week-view/pull/152)

## 0.8.0 - 2021-11-29

- Add precision up to seconds in onGrid callbacks [#149](https://github.com/hoangnm/react-native-week-view/pull/149)

## 0.7.0 - 2021-11-11

- Add isRefreshing and RefreshComponent props [#112](https://github.com/hoangnm/react-native-week-view/pull/112)
- Fix events break when changing hoursInDisplay [#133](https://github.com/hoangnm/react-native-week-view/pull/133)
- Allow to drag-drop event [#143](https://github.com/hoangnm/react-native-week-view/pull/143)
- Support react-native above v0.59.0 only

## 0.6.1 - 2021-08-22

### Fixed

- Fix wrong display when updating `numberOfDays` prop [#126](https://github.com/hoangnm/react-native-week-view/pull/126)

## 0.6.0 - 2021-08-05

### Changed

- Add `fixedHorizontally` prop to display a fixed week [#75](https://github.com/hoangnm/react-native-week-view/pull/75)
- Add `formatTimeLabel` prop [#103](https://github.com/hoangnm/react-native-week-view/pull/103)
- Add `weekStartsOn` prop [#120](https://github.com/hoangnm/react-native-week-view/pull/120)
- Add `TodayHeaderComponent` prop [#113](https://github.com/hoangnm/react-native-week-view/pull/113)

### Fixed

- Fix scroll called twice [#117](https://github.com/hoangnm/react-native-week-view/pull/117)

## 0.5.0 - 2021-05-16

### Changed

- Add `timeStep` prop to control the height of time slot better [#98](https://github.com/hoangnm/react-native-week-view/pull/98)

example/App.js
/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 * @flow strict-local
 */

import React, {useState, useReducer, useRef, useCallback} from 'react';
import {SafeAreaView, StyleSheet, StatusBar, Alert, Text} from 'react-native';

import WeekView, {createFixedWeekDate} from 'react-native-week-view';
import {buildDateCycler, makeBuilder, eventsWithUpdate} from './debug-utils';

const buildEvent = makeBuilder();

const sampleEvents = [
  // Previous week
  buildEvent(-24 * 7 - 5, 2, 'pink'),
  buildEvent(-24 * 7 - 14, 3, 'lightblue'),

  // This week
  buildEvent(0, 2, 'blue', {eventKind: 'block'}),
  buildEvent(1, 3, 'red', {resolveOverlap: 'lane'}),
  buildEvent(-18, 48 + 5, 'green', {eventKind: 'block'}),

  buildEvent(0 - 24, 2, 'blue', {
    allDay: true,
    description: 'Long description is wrapped',
  }),
  buildEvent(1, 3, 'red', {allDay: true}),
  buildEvent(-18, 48 + 5, 'green', {allDay: true}),

  // Next week
  buildEvent(24 * 7, 2, 'magenta'),
  buildEvent(24 * 7 - 48, 3, 'green', {
    style: {
      borderWidth: 5,
    },
    disableDrag: true,
    disablePress: true,
    disableLongPress: true,
  }),
  buildEvent(24 * 7 + 6, 6, 'brown', {resolveOverlap: 'ignore'}),
  buildEvent(24 * 7 + 13, 2, 'lightgreen', {
    allDay: true,
    description: 'Long description is wrapped',
  }),
  buildEvent(24 * 7 - 18, 48 + 5, 'lightgreen', {allDay: true}),

  // Two more weeks
  buildEvent(48 * 7, 2, 'pink'),
  buildEvent(48 * 7 - 54, 4, 'green'),
  buildEvent(48 * 7 + 1, 3, 'lightgreen', {allDay: true}),

  // Many events
  // ...Array.from({length: 1000}, (_, i) =>
  //   buildEvent(24 + i * 5, 2, 'lightblue'),
  // ),
];

const sampleFixedEvents = [
  {
    id: 1,
    description: 'Event 1',
    startDate: createFixedWeekDate('Monday', 12),
    endDate: createFixedWeekDate(1, 14),
    color: 'blue',
  },
  {
    id: 2,
    description: 'Event 2',
    startDate: createFixedWeekDate('wed', 16),
    endDate: createFixedWeekDate(3, 17, 30),
    color: 'red',
  },
];

// For debugging purposes
const showFixedComponent = false;
const INITIAL_EVENTS = showFixedComponent ? sampleFixedEvents : sampleEvents;

const MyRefreshComponent = ({style}) => (
  // eslint-disable-next-line react-native/no-inline-styles
  <Text style={[style, {fontSize: 20, color: 'black'}]}>Loading...</Text>
);

const DRAG_EVENT_CONFIG = null;

const EDIT_EVENT_CONFIG = {
  top: true,
  bottom: true,
  left: true,
  right: true,
};

const PAGE_START_AT = {
  weekday: 1,
};

const onDayPress = (date, formattedDate) => {
  console.log('Day: ', date, formattedDate);
};

const onSwipeNext = d => console.log('Swipe next', d.toDateString());
const onSwipePrev = d => console.log('Swipe prev', d.toDateString());
const onTimeScrolled = d => console.log('Time scrolled', d.toTimeString());

// Use this to manually debug navigate through dates
// eslint-disable-next-line no-unused-vars
const dateCycler = buildDateCycler([
  // // Example:
  // // selectedDate={new Date(2022, 7, 14)}
  // new Date(2022, 7, 20),
  // new Date(2022, 7, 18),
  // new Date(2022, 7, 2),
]);

const App = ({}) => {
  const componentRef = useRef(null);

  const [events, updateEvent] = useReducer(eventsWithUpdate, INITIAL_EVENTS);

  const onDragEvent = useCallback(
    (event, newStartDate, newEndDate) => {
      updateEvent({event, newStartDate, newEndDate});
    },
    [updateEvent],
  );

  const onEditEvent = useCallback(
    (event, newStartDate, newEndDate) => {
      console.log('Editing: ', event.id, newStartDate, newEndDate);
      updateEvent({event, newStartDate, newEndDate});
    },
    [updateEvent],
  );

  const [editingEvent, setEditEvent] = useState(null);

  const handleLongPressEvent = event => {
    if (editingEvent == null) {
      setEditEvent(event.id);
    } else {
      setEditEvent(null);
    }
  };

  const handlePressEvent = event => {
    if (editingEvent != null) {
      setEditEvent(null);
      return;
    }

    const {id, color, startDate, endDate} = event;
    Alert.alert(
      `Event press ${color} - ${id}`,
      `start: ${startDate}\nend: ${endDate}`,
    );
  };

  const handlePressGrid = (event, startHour, date) => {
    if (editingEvent != null) {
      setEditEvent(null);
      return;
    }

    const year = date.getFullYear();
    const month = date.getMonth() + 1; // zero-based
    const day = date.getDate();
    const hour = date.getHours();
    const minutes = date.getMinutes();
    const seconds = date.getSeconds();

    Alert.alert(`${year}-${month}-${day} ${hour}:${minutes}:${seconds}`);
  };

  const onMonthPress = useCallback((date, formattedDate) => {
    // // Debug navigating through dates:
    // if (componentRef && componentRef.current) {
    //   componentRef.current.goToDate(dateCycler.next());
    // }

    console.log('Month: ', date, formattedDate);
  }, []);

  return (
    <>
      <StatusBar barStyle="dark-content" />
      <SafeAreaView style={styles.container}>
        <WeekView
          ref={componentRef}
          events={events}
          selectedDate={new Date()}
          numberOfDays={7}
          pageStartAt={PAGE_START_AT}
          onEventPress={handlePressEvent}
          onEventLongPress={handleLongPressEvent}
          onGridLongPress={handlePressGrid}
          headerStyle={styles.header}
          headerTextStyle={styles.headerText}
          hourTextStyle={styles.hourText}
          eventContainerStyle={styles.eventContainer}
          gridColumnStyle={styles.gridColumn}
          gridRowStyle={styles.gridRow}
          formatDateHeader={showFixedComponent ? 'ddd' : 'ddd DD'}
          hoursInDisplay={12}
          timeStep={60}
          startHour={15}
          fixedHorizontally={showFixedComponent}
          showTitle={!showFixedComponent}
          timesColumnWidth={0.2}
          showNowLine
          onDragEvent={onDragEvent}
          isRefreshing={false}
          RefreshComponent={MyRefreshComponent}
          onDayPress={onDayPress}
          onMonthPress={onMonthPress}
          onTimeScrolled={onTimeScrolled}
          onSwipeNext={onSwipeNext}
          onSwipePrev={onSwipePrev}
          editingEvent={editingEvent}
          onEditEvent={onEditEvent}
          editEventConfig={EDIT_EVENT_CONFIG}
          dragEventConfig={DRAG_EVENT_CONFIG}
          runOnJS={false}
        />
      </SafeAreaView>
    </>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#FFF',
  },
  header: {
    backgroundColor: '#4286f4',
    borderColor: '#fff',
  },
  headerText: {
    color: 'white',
  },
  hourText: {
    color: 'black',
  },
  eventContainer: {
    borderWidth: 1,
    borderColor: 'black',
  },
  gridRow: {
    borderTopWidth: 1,
    borderColor: '#E9EDF0',
  },
  gridColumn: {
    borderLeftWidth: 1,
    borderColor: '#E9EDF0',
  },
});

export default App;
